src/wayland/input_method/input_method_handle.rs

```diff
@@ -83,6 +83,13 @@ impl InputMethodHandle {
         }
     }
 
+    /// Convenience function to close popup surfaces
+    pub fn close_popup(&self) {
+        let inner = self.inner.lock().unwrap();
+        let mut popup = inner.popup.inner.lock().unwrap();
+        popup.surface_role = None;
```

Comment:
Hm. What `input_method` client do you get that behavior with?

Reply:
I get it with mine, but it was also an issue in wlroots I think. But I will need some time to find that again...
Also a summary from the fcitx developer here:
_Currently, few combinations that does work under waylands are:

GNOME-shell + ibus
GNOME-shell directly invokes ibus DBus API to talk to ibus. So the communication looks like
application <--text-input-v3--> GNOME-shell <--ibus DBus API--> ibus

Popup positioning is ensured by GNOME-shell


GNOME-shell + fcitx + kimpanel extension (https://extensions.gnome.org/extension/261/kimpanel/), similar to (1).


Fcitx im module + any compsitor
fcitx im module (only for Gtk/Qt) runs as a plugin within the Gtk/Qt application, and it will render the popup within the application so it looks like
application <--Fcitx DBUS API--> Fcitx<--xdg_popup--> compositor

There's currently no workable zwp_input_method implementation unfortunately.

ibus only has zwp_input_method but without popup support.
sway mainline doesn't yet have input popup support or the "virtual keyboard" part merged, not to mention other wlroots based compositor.
some kwin developer recently present me a screenshot for fcitx + zwp_input_method_v1, so looks like it's a WIP. But I guess zwp_input_method_v1 won't really work with text-input-v3._

I think wlroots now support the whole protocol, and also smithay, this comment is pretty old, but now I understand how they make their popup work. Maybe that would be better than the current special surface role.

Reply:
found it!

https://gitlab.freedesktop.org/wayland/wayland-protocols/-/issues/40#note_835314